### ISSUE TITLE
Neo4j 2.2 Cypher - Cost planner not using indexes for computed values, index hints not working

### ISSUE THREAD

**Issue by bunkat** [neo4j/neo4j#4435]
The COST planner is unable to use indexes for basic MATCH statements, instead choosing to do expensive NodeByLabelScan operations.  This seems to be a problem with using indexes on computed values (cr.teamid) vs with parameters ({p0}) or static strings ("foobar"). I also noticed that the COST planner uses a NodeIndexSeek operation instead of a SchemaIndex operation (for example, when finding the user in the below queries).  It is unclear what the difference between the two is.

The RULE planner uses the SchemaIndex correctly in both instances.

**Query** 

```
  PROFILE
  PLANNER RULE
  MATCH (u:user {id: "foobar"})
  WITH u

  MATCH (u)-[:CONTACT]->(c:contact)-[:CONTACT]->(cr:role)
  WHERE HAS(cr.teamid)
  WITH c, cr

  MATCH (p:project)
  WHERE p.teamid = cr.teamid
  RETURN p
```

**_Result**_

![image](https://cloud.githubusercontent.com/assets/665885/7146829/20d5cf3e-e2ab-11e4-8bc4-ac2f098c5276.png)

**Query** 

```
  PROFILE
  PLANNER COST
  MATCH (u:user {id: "foobar"})
  WITH u

  MATCH (u)-[:CONTACT]->(c:contact)-[:CONTACT]->(cr:role)
  WHERE HAS(cr.teamid)
  WITH c, cr

  MATCH (p:project)
  WHERE p.teamid = cr.teamid
  RETURN p
```

**_Result**_

![image](https://cloud.githubusercontent.com/assets/665885/7146851/502734da-e2ab-11e4-85a8-6a0044358254.png)

**Query** 

```
  PROFILE
  PLANNER COST
  MATCH (u:user {id: "foobar"})
  WITH u

  MATCH (u)-[:CONTACT]->(c:contact)-[:CONTACT]->(cr:role)
  WHERE HAS(cr.teamid)
  WITH c, cr

  MATCH (p:project)
  USING INDEX p:project(teamid)
  WHERE p.teamid IN [cr.teamid]
  RETURN p
```

**_Result**_

![image](https://cloud.githubusercontent.com/assets/665885/7146884/96cce45c-e2ab-11e4-9c44-5edfa577511e.png)

**Comment by craigtaverner** [comment]
The short answer is that you are right that the cost planner is not optimised for this case where the index lookup would be based on a dynamic or computed value. Another way of wording it is that it is not optimised for a 'foreign key join'. In this case you could have used a relationship to connect the (cr:role) and (p:project) nodes. The connection is by the teamid, and you would perhaps use a (cr)-[:team]-(p) relationship. If the value of the teamid is important, you could put a property on this relationship or make an intermediate (t:team) node.

Getting back to your query though, it should be possible to teach the cost model to consider using an index for even this case. I would think of that as a performance optimization. However, I'm curious to know what the performance comparison of the RULE and COST queries was? Was RULE much faster than COST? Even though COST did not use the schema index, it is possible that the NodeByLabelScan would have been fast enough anyway if the number of projects is small.

One thing that does look like a bug here is the error message for the index hint. It seems to claim that no such index exists, which is not the case. So we seem to have a bug and a feature request:
- bug - wrong error message on index hint
- feature request - performance optimization to use schema index on dynamic property (Apply/NodeIndexSeek instead of CartesianProduct/NodeByLabelScan)

**Comment by bunkat** [comment]
For the queries I posted:

```
  PROFILE
  MATCH (u:user {id: "foobar"})
  WITH u

  MATCH (u)-[:CONTACT]->(c:contact)-[:CONTACT]->(cr:role)
  WHERE HAS(cr.teamid)
  WITH c, cr

  MATCH (p:project)
  WHERE p.teamid = cr.teamid
  RETURN p

  Cypher version: CYPHER 2.2, planner: RULE. 224 total db hits in 87 ms (10 run avg).
  Cypher version: CYPHER 2.2, planner: COST. 1098 total db hits in 117 ms (10 run avg).
```

The COST planner takes 4.9x db hits and 1.34x as long to execute.  This was just a simple demonstrative query example with a test database of 150 projects.  Other queries take 4x to 10x longer to execute when using the COST planner and all of them will just continue to get slower over time due to the NodeByLabelScan operations.

While I could add more and more relationships to try and overcome these limitations it becomes very hard to manage.  Especially since trying to avoid this type of foreign key join would mean that every child object would need to have an additional 10+ relationships so that I could always quickly get to the associated team or project (since there are role hierarchies 5 deep for both teams and projects that would all need these new connections).  Plus, none of this is needed for the RULE planner.

In general it would be helpful to have more information on when the COST planner is expected to be the better option.  I found the following quote in the blog post but it makes no mention on how the indices for the COST planner need to be used or other drawbacks:

> So, for which types of queries does Ronja bring about the best improvements?
> We have conducted extensive benchmarking tests, and, empirically, queries
> containing multiple nodes that can be reached via indices and/or pattern predicates
> show the most improvement. 

It's unfortunate, but dropping in Neo4j 2.2 with default behavior has introduced a lot of new performance bottlenecks that made the database unfit for production use in my case.  At this point, I just always force the RULE planner and see a lot better and more consistent performance.

**Comment by jexp** [comment]
Hey Bill, sorry you ran into this, it's a bug which will be fixed.

Currently computed expressions are not taken into account for index lookups in the cost planner.

But you can force it to use an index, I'd be interested in the comparison between cost and rule then:

```
  PROFILE
  MATCH (u:user {id: "foobar"})
  WITH u

  MATCH (u)-[:CONTACT]->(c:contact)-[:CONTACT]->(cr:role)
  WHERE HAS(cr.teamid)
  WITH c, cr

  MATCH (p:project)
  using index p:project(teamid)
  WHERE p.teamid = cr.teamid
  RETURN p
```

**Comment by bunkat** [comment]
That was the other bug that I hit.  You can't use index hints with the COST planner.

**Query as suggested fails to compile**

```
  PROFILE
  MATCH (u:user {id: "foobar"})
  WITH u

  MATCH (u)-[:CONTACT]->(c:contact)-[:CONTACT]->(cr:role)
  WHERE HAS(cr.teamid)
  WITH c, cr

  MATCH (p:project)
  using index p:project(teamid)
  WHERE p.teamid = cr.teamid
  RETURN p
```

> Cannot use index hint in this context. Index hints require using a simple equality comparison or IN  condition in WHERE (either directly or as part of a top-level AND). Note that the label and property  comparison must be specified on a non-optional node (line 9, column 3 (offset: 164))
>  "  using index p:project(teamid)"

**Switching to an IN expression, query fails to find index**

```
  PROFILE
  MATCH (u:user {id: "foobar"})
  WITH u

  MATCH (u)-[:CONTACT]->(c:contact)-[:CONTACT]->(cr:role)
  WHERE HAS(cr.teamid)
  WITH c, cr

  MATCH (p:project)
  using index p:project(teamid)
  WHERE p.teamid IN [cr.teamid]
  RETURN p
```

> No such index found.
> Label: `project`
> Property name: `teamid`

```
Indexes
  ....
  ON :project(teamid)        ONLINE  
```

**Using the RULE planner instead, indexes work as expected**

```
  PROFILE
  PLANNER RULE
  MATCH (u:user {id: "foobar"})
  WITH u

  MATCH (u)-[:CONTACT]->(c:contact)-[:CONTACT]->(cr:role)
  WHERE HAS(cr.teamid)
  WITH c, cr

  MATCH (p:project)
  using index p:project(teamid)
  WHERE p.teamid IN [cr.teamid]
  RETURN p
```

> Cypher version: CYPHER 2.2, planner: RULE. 224 total db hits in 87 ms.

By the way, this is actually a breaking change since this type of query will attempt to use the COST planner by default and then fail on `using index` throwing an error.  Suddenly queries that worked fine in 2.1.8 will be broken when the database is upgraded to 2.2.1.

**Comment by jexp** [comment]
Sorry for that, really frustrating.

```
PROFILE
  MATCH (u:user {id: "foobar"})
  WITH u

  MATCH (u)-[:CONTACT]->(c:contact)-[:CONTACT]->(cr:role)
  WHERE HAS(cr.teamid)
  WITH c, cr, cr.teamid as teamid

  MATCH (p:project)
  using index p:project(teamid)
  WHERE p.teamid = teamid
  RETURN p
```

**Comment by bunkat** [comment]
>  Cypher version: CYPHER 2.2, planner: RULE. 224 total db hits in 94 ms (10 run avg).

Using the index hint (did not know that node properties weren't allowed...) brings the performance of the COST planner in line with the RULE planner for this example.  It still isn't faster, but they are basically equivalent.

**Comment by jexp** [comment]
Node properties and other expressions actually are allowed, it has to be fixed.

Ya for this simple query I doubt that the performance would be much different. Feel free to share the profile output for both planners.

**Comment by bunkat** [comment]
**RULE**

![image](https://cloud.githubusercontent.com/assets/665885/7171147/9b6c0554-e389-11e4-901b-844355ac8a8e.png)

**COST**

The final Apply and Projection steps look a little funny, but I guess that is just the COST planner doing the join?

![image](https://cloud.githubusercontent.com/assets/665885/7171156/b5283a3a-e389-11e4-9fac-d1bc7bcf7fae.png)

**Comment by jjaderberg** [comment]
I see a similar problem where the cost based planner fails with index hints. In my case the values are string literals, not properties or expressions, so not sure if it's the same problem.

Executing the following query with the cost planner in Cypher 2.2 gives various exceptions with the message `The current planner cannot satisfy all hints in the query, please try removing hints or try with another planner`.

```
MATCH (a:Tag)-[:RELATED*1..2]-t-[:RELATED*1..2]-(b:Tag), t-[:RELATED*1..2]-(c:Tag)
USING INDEX a:Tag(name)
USING INDEX b:Tag(name)
USING INDEX c:Tag(name)
WHERE a.name = "A" AND b.name = "B" AND c.name = "C"
RETURN DISTINCT t
```

In  the shell I get a `QueryExecutionKernelException`.

![shell-22-cost](https://cloud.githubusercontent.com/assets/6028968/10395723/c1cf15f6-6e9e-11e5-8dda-487334d5620a.png)

The browser gives a `Neo.ClientError.Schema.NoSuchIndex`, but the message is the same.

![browser-22-cost](https://cloud.githubusercontent.com/assets/6028968/10395720/c1c98eec-6e9e-11e5-9b78-b8f0bdcace85.png)

I set up a [sample graph](http://console.neo4j.org/r/2mi9fv) in a Rabbithole console (the data is there but you have to `CREATE CONSTRAINT ON (tag:Tag) ASSERT tag.name IS UNIQUE` before testing the query). In the console I get a `HintException`, but again the message is the same.

![console-22-cost](https://cloud.githubusercontent.com/assets/6028968/10395721/c1ca8586-6e9e-11e5-91b1-8263bdfe710a.png)

The same query works fine with the cost planner in the 2.3 compiler in the console.

![console-23-cost](https://cloud.githubusercontent.com/assets/6028968/10395722/c1cd6300-6e9e-11e5-99e2-6e1f1e119af3.png)

It also works with Cypher 2.2 if I force the rule planner.

![console-22-rule](https://cloud.githubusercontent.com/assets/6028968/10395724/c1cf8446-6e9e-11e5-89f8-e86f97596cdb.png)

**Comment by Mats-SX** [comment]
@jjaderberg What you're seeing is a constraint with our first version of cost-based planner, which we call `greedy`. The `greedy` planner only supports one index hint. If you want to use more than one in 2.2, you'll have to specify the planner to `idp`. In 2.3, `idp` is the default for cost-based planner, so multiple hints will work from the get-go.

**Comment by Mats-SX** [comment]
@bunkat Since the PR https://github.com/neo4j/neo4j/pull/6162 was merged into 3.0, this problem should be fixed. Is it possible for you to verify this on your data?

**Comment by Mats-SX** [comment]
Closing this as fixed. Please reopen in case it can be reproduced.
